xla/service/gpu/gemm_rewriter.cc

```diff
@@ -1285,24 +1444,36 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
 
   // Adds a scalar DAmax return value to an FP8 GEMM.
   absl::Status F8AddDAmax(HloInstruction *instr, HloInstruction *existing_gemm,
-                          HloInstruction *reduce_damax) {
+                          HloInstruction *reduce_damax,
+                          HloInstruction *get_tuple_element) {
     // Change the output shape of the Custom Call to tuple(D, DAmax).
     Shape damax_shape = ShapeUtil::MakeScalarShape(F32);
-    Shape tuple_shape =
-        ShapeUtil::MakeTupleShape({instr->shape(), damax_shape});
+
+    std::vector<Shape> tuple_shape_vec = {instr->shape(), damax_shape};
+    if (get_tuple_element) {
+      tuple_shape_vec.insert(tuple_shape_vec.begin() + 1,
+                             existing_gemm->shape().tuple_shapes(1));
+    }
+    Shape tuple_shape = ShapeUtil::MakeTupleShape(tuple_shape_vec);
     HloInstruction *gemm_and_damax =
         instr->AddInstruction(existing_gemm->CloneWithNewShape(tuple_shape));
 
     // Obtain D and DAmax separately from the output tuple.
     HloInstruction *d =
         instr->AddInstruction(HloInstruction::CreateGetTupleElement(
             instr->shape(), gemm_and_damax, 0));
-    HloInstruction *damax = instr->AddInstruction(
-        HloInstruction::CreateGetTupleElement(damax_shape, gemm_and_damax, 1));
+    HloInstruction *damax =
+        instr->AddInstruction(HloInstruction::CreateGetTupleElement(
+            damax_shape, gemm_and_damax, (get_tuple_element ? 2 : 1)));
 
     // Convert DAmax from FP32 to the requested type and elide reduce.
     HloInstruction *damax_converted = instr->AddInstruction(
         HloInstruction::CreateConvert(reduce_damax->shape(), damax));
+
+    if (get_tuple_element) {
+      TF_RETURN_IF_ERROR(instr->parent()->ReplaceInstructionWithDifferentShape(
+          existing_gemm, gemm_and_damax));
```

Comment:
I know I suggested this before, but I actually prefer not to replace existing_gemm with gemm_and_damax. This reason is that you change the dtype of the first output of existing_gemm, so if there are any users of this output, this replacement will fail. 

The old logic you have I don't think is correct anymore either since even gemms without a bitmask can output a tuple. Instead, I would check for users for each of the outputs of gemm (other than the first output and the amax output) and replace them a new GTE instruction